L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+      bool valid(true);
+      // having already maximum number of added layers
+      if (hitPattern.count() == setup_->kfMaxLayers())
+        valid = false;
+      // not enough layers remain after skipping
+      if (hitPattern.count() + track->hitPattern().count(layer + 1, setup_->numLayers()) < setup_->kfMaxLayers())
+        valid = false;
+      if (valid) {
+        // pick next stub on next populated layer
+        for (int nextLayer = layer_ + 1; nextLayer < setup_->numLayers(); nextLayer++) {
+          if (track->hitPattern(nextLayer)) {
+            stubNext = track->layerStub(nextLayer);
+            break;
+          }
+        }
+      }
+    }
+    if (stubNext) {
+      states_.emplace_back(state, stubNext);
+      state = &states_.back();
+    } else
+      state = nullptr;
+  }
+
+  // best state selection
+  void KalmanFilter::accumulator(deque<State*>& stream) {
+    // accumulator delivers contigious stream of best state per track
+    // remove gaps and not final states
+    stream.erase(remove_if(stream.begin(), stream.end(), [this](State* state){ return !state || state->hitPattern().count() < setup_->kfMaxLayers(); }), stream.end());
+    /*auto toLessPS = [this](State* state) {
+      int nPS(0);
+      for (const StubKF& stub : state->stubs())
+        if (setup_->psModule(stub.ttStubRef()))
+          nPS++;
+      return nPS < 2;
+    };
+    stream.erase(remove_if(stream.begin(), stream.end(), toLessPS), stream.end());*/
+    // update chi2
+    for (State* state : stream)
+      state->finish();
+    // sort in number of skipped layers
+    auto lessSkippedLayers = [](State* lhs, State* rhs) { return lhs->numSkippedLayers() < rhs->numSkippedLayers(); };
+    stable_sort(stream.begin(), stream.end(), lessSkippedLayers);
+    // sort in number of consistent stubs
+    auto moreConsistentLayers = [](State* lhs, State* rhs) { return lhs->numConsistentLayers() > rhs->numConsistentLayers(); };
+    stable_sort(stream.begin(), stream.end(), moreConsistentLayers);
+    // sort in track id
+    stable_sort(stream.begin(), stream.end(), [](State* lhs, State* rhs){ return lhs->trackId() < rhs->trackId(); });
+    // keep first state (best due to previous sorts) per track id
+    stream.erase(unique(stream.begin(), stream.end(), [](State* lhs, State* rhs){ return lhs->track() == rhs->track(); }), stream.end());
+    //for (State* state : stream)
+      //cout << state->x0() << " " << state->x1() << " " << dataFormats_->chosenRofPhi() << endl;
+  }
+
+  // updates state
+  void KalmanFilter::update(State*& state) {
+    // R-Z plane
+    const double H12 = H12_->digi(state->H12());
+    const double m1 = m1_->digi(state->m1());
+    const double v1 = v1_->digi(state->v1());
+    double x2 = x2_->digi(state->x2());
+    double x3 = x3_->digi(state->x3());
+    double C22 = C22_->digi(state->C22());
+    double C23 = C23_->digi(state->C23());
+    double C33 = C33_->digi(state->C33());
+    v1_->updateRangeActual(v1);
```

Comment:
Do the calls to updateRangeActual() have no effect on the KF, and are only used in the endJob to recommend changes in the ranges? If so, add comment saying this.

Reply:
done